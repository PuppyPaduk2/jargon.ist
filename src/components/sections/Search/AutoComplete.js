// @flow

import React, { PureComponent } from 'react';
import { navigateTo } from 'gatsby-link';
import { connectAutoComplete } from 'react-instantsearch/connectors';
import Autosuggest from 'react-autosuggest';

import { type Jargon } from '../../../types/jargon';
import { type Tag } from '../../../types/tag';

type IndexType = 'jargons' | 'tags';
type Hit = (Jargon | Tag) & { index: IndexType };

type Section = {
  index: IndexType,
  hits: Array<Jargon | Tag>,
};

type Props = {
  hits: Array<Jargon | Tag>,
  currentRefinement: string,
  refine: string => void,
};

export class AutoCompleteWrapper extends PureComponent<Props> {
  titles: { [string]: string };

  constructor(props: Props) {
    super(props);

    this.titles = {
      jargons: 'Jargonlar',
      tags: 'Konular',
    };
  }

  handleChange = (event: any, { newValue, method }: { newValue: string, method: string }) => {
    if (['click', 'enter'].indexOf(method) > -1) {
      navigateTo(newValue);
    }
  };

  handleSuggestionsFetchRequested = ({ value }: any) => {
    this.props.refine(value);
  };

  handleSuggestionsClearRequested = () => {
    this.props.refine('');
  };

  getSuggestionValue = (hit: Hit) => {
    const prefix = hit.index === 'jargons' ? '' : '/k';

    return `${prefix}/${hit.slug}`;
  };

  getSectionSuggestions = (section: Section) =>
    section.hits.map(hit => ({ ...hit, index: section.index }));

  renderSuggestion = (hit: Hit) => hit.title;

  renderSectionTitle = (section: Section) =>
    section.hits.length > 0 && (
      <h4 className="c-search__suggestions-section-title">{this.titles[section.index]}</h4>
    );

  render() {
    const { hits, currentRefinement } = this.props;

    return (
      <Autosuggest
        suggestions={hits}
        multiSection
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        getSuggestionValue={this.getSuggestionValue}
        getSectionSuggestions={this.getSectionSuggestions}
        renderSuggestion={this.renderSuggestion}
        inputProps={{
          placeholder: 'Jargon veya konu ara',
          value: currentRefinement,
          onChange: this.handleChange,
        }}
        renderSectionTitle={this.renderSectionTitle}
        theme={{
          input: 'c-search__input',
          suggestionsContainer: 'c-search__suggestions-container',
          suggestionsContainerOpen: 'c-search__suggestions-container--open',
          suggestionsList: 'c-search__suggestions-list',
          suggestion: 'c-search__suggestion',
          suggestionHighlighted: 'c-search__suggestion--highlighted',
        }}
      />
    );
  }
}

export default connectAutoComplete(AutoCompleteWrapper);
