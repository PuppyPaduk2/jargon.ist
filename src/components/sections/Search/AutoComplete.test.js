// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { navigateTo } from 'gatsby-link';

import { AutoCompleteWrapper } from './AutoComplete';

jest.mock('gatsby-link');

beforeEach(() => {
  navigateTo.mockClear();
});

it('renders correctly', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders sections title if has some hits', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  expect(wrapper.instance().renderSectionTitle({ hits: [] })).toBeFalsy();
  expect(wrapper.instance().renderSectionTitle({ hits: [{}] })).toBeTruthy();
});

it('adds index to hits for suggestion sections', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  const section = {
    index: 'index1',
    hits: [{ title: 'title 1', slug: 'title-1' }],
  };

  const sectionSuggestions = [
    {
      title: 'title 1',
      slug: 'title-1',
      index: 'index1',
    },
  ];

  expect(wrapper.instance().getSectionSuggestions(section)).toEqual(sectionSuggestions);
});

it('navigates to related detail when click to result', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  wrapper.instance().handleChange(undefined, { newValue: 'hello', method: 'click' });
  expect(navigateTo).toHaveBeenCalled();
});

it('does nothing on handleChange if event is not click or enter', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  wrapper.instance().handleChange(undefined, { newValue: 'hello', method: 'type' });
  expect(navigateTo).not.toHaveBeenCalled();
});

it('creates proper links for details', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  const jargonHit = {
    index: 'jargons',
    title: 'jargon 1',
    slug: 'jargon-1',
  };

  const tagHit = {
    index: 'tags',
    title: 'tag 1',
    slug: 'tag-1',
  };

  expect(wrapper.instance().getSuggestionValue(jargonHit)).toEqual(`/${jargonHit.slug}`);
  expect(wrapper.instance().getSuggestionValue(tagHit)).toEqual(`/k/${tagHit.slug}`);
});

it('renders title as suggestion item', () => {
  const wrapper = shallow(<AutoCompleteWrapper hits={[]} currentRefinement="" refine={() => {}} />);

  const hit = {
    index: 'tags',
    title: 'tag 1',
    slug: 'tag-1',
  };

  expect(wrapper.instance().renderSuggestion(hit)).toEqual(hit.title);
});
