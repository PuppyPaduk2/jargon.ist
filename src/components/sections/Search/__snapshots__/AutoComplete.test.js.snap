// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Autosuggest
  alwaysRenderSuggestions={false}
  focusInputOnSuggestionClick={true}
  getSectionSuggestions={[Function]}
  getSuggestionValue={[Function]}
  highlightFirstSuggestion={false}
  id="1"
  inputProps={
    Object {
      "autoCapitalize": "off",
      "autoComplete": "off",
      "autoCorrect": "off",
      "onChange": [Function],
      "placeholder": "Jargon veya konu ara",
      "type": "search",
      "value": "",
    }
  }
  multiSection={true}
  onSuggestionsClearRequested={[Function]}
  onSuggestionsFetchRequested={[Function]}
  renderSectionTitle={[Function]}
  renderSuggestion={[Function]}
  renderSuggestionsContainer={[Function]}
  shouldRenderSuggestions={[Function]}
  suggestions={Array []}
  theme={
    Object {
      "input": "c-search__input",
      "suggestion": "c-search__suggestion",
      "suggestionHighlighted": "c-search__suggestion--highlighted",
      "suggestionsContainer": "c-search__suggestions-container",
      "suggestionsContainerOpen": "c-search__suggestions-container--open",
      "suggestionsList": "c-search__suggestions-list",
    }
  }
/>
`;
